_posts/2021-03-23-keccak-optimizer-bug.md

```diff
@@ -0,0 +1,89 @@
+---
+layout: post
+published: true
+title:  "Solidity Keccak Caching bug"
+date:   2021-03-23
+author: Solidity Team
+category: Security Alerts
+---
+
+On March 20, 2021, a bug in Solidity's bytecode optimizer was found. The bug is fixed with [version
+0.8.3](https://github.com/ethereum/solidity/releases/tag/v0.8.3) released on March 23, 2021. **The
+bug is present in all prior versions of Solidity.**
+
+We assigned the bug a severity level of "medium".
+
+## Technical Details of the Bug
+
+**Summary**: The bytecode optimizer incorrectly re-used previously evaluated Keccak-256 hashes. You
+are unlikely to be affected if you do not compute Keccak-256 hashes in inline assembly.
+
+Solidity's bytecode optimizer has a step that can compute Keccak-256 hashes, if the contents of the
+memory are known during compilation time. This step also has a mechanism to determine that two
+Keccak-256 hashes are equal even if the values in memory are not known during compile time. This
+mechanism had a bug where, Keccak-256 of the same memory content, but different sizes were
```

Comment:
To be sure, the bug is in the mechanism which infers keccak equality by memory location only. This mechanism is used in two scenarios, right?
  - When memory content is compile time constant and identical across two keccak calls
  - When memory content is not compile-time constant but identical across two keccak calls

The wording of the sentence gives the impression that the bug is only in scenario (2)

Reply:
Hopefully the new updated example would avoid this confusion.